sqlserver/datadog_checks/sqlserver/schemas.py

```diff
@@ -0,0 +1,369 @@
+    PARTITIONS_QUERY,
+    SCHEMA_QUERY,
+    STATIC_INFO_ENGINE_EDITION,
+    STATIC_INFO_VERSION,
+    TABLES_IN_SCHEMA_QUERY,
+)
+from datadog_checks.sqlserver.utils import execute_query_output_result_as_a_dict, get_list_chunks
+
+
+class SubmitData:
+    MAX_COLUMN_COUNT = 10_000
+
+    # TBD - REDAPL has a 3MB limit per resource
+    # If a column payload is ~ 10bytes : name, type, default , if nullable nullable
+    # then the limit should be only 25_000.
+
+    MAX_TOTAL_COLUMN_COUNT = 100_000
+
+    def __init__(self, submit_data_function, base_event, logger):
+        self._submit_to_agent_queue = submit_data_function
+        self._base_event = base_event
+        self._log = logger
+
+        self._columns_count = 0
+        self._total_columns_count = 0
+        self.db_to_schemas = {}  # dbname : { id : schema }
+        self.db_info = {}  # name to info
+
+    def set_base_event_data(self, hostname, tags, cloud_metadata, dbms_version):
+        self._base_event["host"] = hostname
+        self._base_event["tags"] = tags
+        self._base_event["cloud_metadata"] = cloud_metadata
+        self._base_event["dbms_version"] = dbms_version
+
+    def reset(self):
+        self._columns_count = 0
+        self._total_columns_count = 0
+        self.db_to_schemas = {}
+        self.db_info = {}
+
+    def store_db_info(self, db_name, db_info):
+        self.db_info[db_name] = db_info
+
+    def store(self, db_name, schema, tables, columns_count):
+        self._columns_count += columns_count
+        self._total_columns_count += columns_count
+        schemas = self.db_to_schemas.setdefault(db_name, {})
+        if schema["id"] in schemas:
+            known_tables = schemas[schema["id"]].setdefault("tables", [])
+            known_tables = known_tables + tables
+        else:
+            schemas[schema["id"]] = copy.deepcopy(schema)
+            schemas[schema["id"]]["tables"] = tables
+        if self._columns_count > self.MAX_COLUMN_COUNT:
+            self._submit()
+
+    def exceeded_total_columns_number(self):
+        return self._total_columns_count > self.MAX_TOTAL_COLUMN_COUNT
+
+    #NOTE: DB with no schemas is never submitted
+    def submit(self):
+        if not bool(self.db_to_schemas):
+            return
+        self._columns_count = 0
+        event = {**self._base_event, "metadata": [], "timestamp": time.time() * 1000}
+        for db, schemas_by_id in self.db_to_schemas.items():
+            db_info = {}
+            if db not in self.db_info:
+                self._log.error("Couldn't find database info for %s", db)
+                db_info["name"] = db
+            else:
+                db_info = self.db_info[db]
+            event["metadata"] = event["metadata"] + [{**(db_info), "schemas": list(schemas_by_id.values())}]
+        json_event = json.dumps(event, default=default_json_event_encoding)
+        self._log.debug("Reporting the following payload for schema collection: {}".format(json_event))
```

Comment:
I don't think we want to log the entire schema payload here. If a customer enables debug logs for some other reason, this will absolutely flood their agent logs

Reply:
true, let me shorten it.